_posts/cqrs-and-es.md

```diff
@@ -0,0 +1,63 @@
+
+Teams could embrace CQRS as their approach to designing most, if not all, the query/command interactions; they could even elect CQRS as a top-level architecture or a general approach to system design. If that's the case, and the implementation requires event sourcing, they're also bringing ES everywhere in the system, even where it makes no sense or does more harm than good.
+
+## Back to basics
+
+Let's take a step back to understand the goal here. As software developers, we should strive to design maintainable and evolvable software. Unless what we're releasing will never be touched again, sooner or later, we'll be faced with requests to fix something that doesn't work as planned or to evolve or add features. If the overall design is terrible, it'll be difficult to fulfill those requests.
+
+CQRS, as an instance of the Single Responsibility Principle (SRP), greatly helps design more straightforward software to maintain and evolve. And that's it. It doesn't dictate that we need to use event sourcing, employ different databases, or use messages and queues to propagate changes from the write models to the read ones.
+
+CQRS is about drawing a thick line separating the system's writing and reading pipes. It tells us not to mix write and read models but to keep them well isolated because they have different purposes.
+
+## And in practice?
+
+Translating the presented principle in practice means, for example, that we should not be issuing SQL queries directly to a table. Instead, we could use stored procedures to write data (commands) and views to read data (queries).
+
+If, in our codebase, we use an ORM, Entity Framework, for example, we could have different data contexts and different models for the command and query parts of the system. At this point, we might realize that the queries-related models and data contexts are always read-only, and [Dapper](https://github.com/DapperLib/Dapper) is a better lightweight alternative to Entity Framework for the queries.
+
+If the system exposes API endpoints, querying endpoints should be separated from write endpoints. Or better yet, clients use HTTP endpoints to read data (queries), queues and messages to request changes (commands).
+
+In all the presented examples, and in the many more we could think about, the stress is not on the technology—that's only supporting the example—it's clearly on differentiating queries and commands.
+
+## Conclusion
+
+Carefully understanding the differences between an architectural style and a persistence tactic is critical to not mixing up apples and oranges where and when it doesn't make sense. As discussed, CQRS doesn't imply event sourcing. They are designed to address different concerns. Sometimes, we can use them in conjunction. At the same time, we also want to design systems so that we can choose when and where to use them.
```

Comment:
```suggestion
Carefully understanding the differences between an architectural style and a persistence tactic is critical to not mixing apples and oranges where and when it doesn't make sense. As discussed, CQRS doesn't imply Event Sourcing. They are designed to address different concerns. Sometimes, we can use them in conjunction. At the same time, we also want to design systems so that we can choose when and where to use them.
```